compiler/src/dotty/tools/dotc/transform/init/Checking.scala

```diff
@@ -0,0 +1,423 @@
+        indentCount += 1
+        val pos = tree.sourcePos
+        val line = "[ " + pos.source.file.name + ":" + (pos.line + 1) + " ]"
+        if (last != line)
+          sb.append(
+            if (pos.source.exists)
+              i"${ " " * indentCount }-> ${pos.lineContent.trim}\t$line\n"
+            else
+              i"${tree.show}\n"
+          )
+        last = line
+      }
+      sb.toString
+    }
+  }
+
+  private implicit def theEnv(implicit state: State): Env = state.env
+
+  /** Check that the given concrete class may be initialized safely
+   *
+   *  It assumes that all definitions are properly summarized before-hand.
+   *  However, summarization can be done lazily on-demand to improve
+   *  performance.
+   */
+  def checkClassBody(cdef: TypeDef)(implicit state: State): Unit = traceOp("checking " + cdef.symbol.show, init) {
+    val cls = cdef.symbol.asClass
+    val tpl = cdef.rhs.asInstanceOf[Template]
+
+    // mark current class as initialized, required for linearization
+    state.parentsInited += cls
+
+    def checkClassBodyStat(tree: Tree)(implicit ctx: Context): Unit = traceOp("checking " + tree.show, init) {
+      tree match {
+        case vdef : ValDef =>
+          val (pots, effs) = Summarization.analyze(vdef.rhs)(theEnv.withOwner(vdef.symbol))
+          theEnv.summaryOf(cls).cacheFor(vdef.symbol, (pots, effs))
+          if (!vdef.symbol.is(Flags.Lazy)) {
+            checkEffects(effs)
+            traceIndented(vdef.symbol.show + " initialized", init)
+            state.fieldsInited += vdef.symbol
+          }
+
+        case tree =>
+          val (_, effs) = Summarization.analyze(tree)
+          checkEffects(effs)
+      }
+    }
+
+    def checkEffects(effs: Effects): Unit = {
+      val rebased = Effects.asSeenFrom(effs, ThisRef(state.thisClass)(null), cls, Potentials.empty)
+      rebased.foreach { check(_) }
+    }
+
+    // check parent calls : follows linearization ordering
+    // see spec 5.1 about "Template Evaluation".
+    // https://www.scala-lang.org/files/archive/spec/2.13/05-classes-and-objects.html
+
+    def checkCtor(ctor: Symbol, tp: Type, source: Tree)(implicit ctx: Context): Unit = {
+      val cls = ctor.owner
+      val classDef = cls.defTree
+      if (!classDef.isEmpty) {
+        if (ctor.isPrimaryConstructor) checkClassBody(classDef.asInstanceOf[TypeDef])
+        else checkSecondaryConstructor(ctor)
+      }
+      else if (!cls.isOneOf(Flags.EffectivelyOpenFlags))
+        ctx.warning("Inheriting non-open class may cause initialization errors", source.sourcePos)
+    }
+
+    cls.paramAccessors.foreach { acc =>
+      if (!acc.is(Flags.Method)) {
+        traceIndented(acc.show + " initialized", init)
+        state.fieldsInited += acc
+      }
+    }
+
+    tpl.parents.foreach {
+      case tree @ Block(_, parent) =>
+        val (ctor, _, _) = decomposeCall(parent)
+        checkCtor(ctor.symbol, parent.tpe, tree)
+
+      case tree @ Apply(Block(_, parent), _) =>
```

Comment:
It comes from enabling the flag for the whole test set, where we may encounter such cases. Indeed, it would be nice to minimize a test case and add a comment here.